Is31pixelbuf

gamblor21 opened the pull request:
Added new functions to the is31fl3741 module similar to `neopixel_write` but for is31fl3741 LEDs. Included an initialize function as well. These two functions are used with a yet to be submitted python library to use with with the adafruit_pixelbuf class.

Also includes a small bug fix to the existing library and changes the map from `list` to `tuple` upon advice I was given. There are no libraries released that were lists so best to do so now.

tannewt commented on shared-bindings/is31fl3741/__init__.c:
```diff
@@ -24,16 +24,89 @@
+
+//| """Low-level is31fl3741 implementation
+//|
+//| The `is31fl3741_write` module contains a helper method to write out bytes in
+//| over the I2C bus."""
```
I don't think you need this because the module is `is31fl3741` and has more than the helper function.

tannewt commented on shared-bindings/is31fl3741/__init__.c:
```diff
@@ -24,16 +24,89 @@
+//|     :param ~busio.I2C i2c: the I2C bus to output with
+//|     :param ~int addr: the I2C address of the IS31FL3741 device"""
+//|     ...
+STATIC mp_obj_t is31fl3741_is31fl3741_init(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
```
This feels weird to me because it is initing something without creating an object for it.

gamblor21 replied:
It is initing the driver chip but thinking about this some more, I think this portion can easily be moved to python. The init code isn't speed sensitive.

tannewt reviewed (commented):
Why does this need to be done in C? Can't write be done from Python alone?

gamblor21 commented:
> Why does this need to be done in C? Can't write be done from Python alone?

I did it in C as I thought it would need it for speed (especially the mapping) and for coordination with the displayio work I did. I will do a test of C vs Python to see if it is required. Now that I understand how all the parts work doing it in Python may make sense.

tannewt commented:
I didn't see any coordination here. Did I miss it?

If it does need to be coordinated, then I'd expect it to be part of the existing object.

It'd also be helpful if you'd link to the prototype library.

Thanks!

gamblor21 commented:
The python library (not yet merged upstream) https://github.com/gamblor21/Adafruit_CircuitPython_IS31FL3741/tree/native_changes

I added a `is31fl3741_pixelbuf.py` that is a subclass of pixelbuf the same way Neopixel is. Also `led_glasses_map.py` that provides the pre-defined mapping coordinates that a user will need if they're using the glasses.

I did some testing to see if the `is31fl3741_write` makes sense in C vs python, all this was modeled after the `neopixel_write` submodule. Doing the mapping of the LEDs from pixels on the glasses is faster in C (just over twice as fast). For the test I just changed every pixel color repeatedly. Just like the `neopixel_write` function this is meant for the entire display being written, same as the existing buffered functionality.

It may be easiest to discuss this at the next CP meeting.

tannewt commented:
@gamblor21 Sorry I wasn't at this weeks meeting. I took the day off. What did you conclude there? 2x speed-up seems worth it.

gamblor21 commented:
> @gamblor21 Sorry I wasn't at this weeks meeting. I took the day off. What did you conclude there? 2x speed-up seems worth it.

I was checking with Kattni on the library end of things which looks good (it won't break the existing library on smaller boards). I'll clean up the code I have with the comments above and also submit the PR to the library so they can be looked at together.

tannewt commented:
I see you pushed the library PR. Let me know when you'd like me to look at this and I will. Thanks!

gamblor21 commented:
> I see you pushed the library PR. Let me know when you'd like me to look at this and I will. Thanks!

Both PRs are passing checks and I did a final check on my hardware here so should be good for reviews.

tannewt commented on shared-bindings/is31fl3741/__init__.c:
```diff
@@ -24,16 +24,59 @@
+#include "shared-bindings/busio/I2C.h"
 #include "shared-bindings/is31fl3741/IS31FL3741.h"
 
+//| def is31fl3741_write(i2c: busio.I2C, addr: int, mapping: Tuple[int, ...], buf: ReadableBuffer) -> None:
```
I think this should be moved into the IS31FL3741 object. Providing the same i2c bus and device id to three objects in the lib seems weird. I'll post a suggestion to a library example.

tannewt commented on shared-bindings/is31fl3741/FrameBuffer.c:
```diff
@@ -0,0 +1,303 @@
+//|     """Displays an in-memory framebuffer to a IS31FL3741 drive display."""
+//|
+
+//|     def __init__(self, *, width: int) -> None:
```
Update this please.

tannewt commented on shared-bindings/is31fl3741/IS31FL3741.c:
```diff
@@ -32,82 +32,35 @@
+//|     """Driver for a IS31FL3741 device."""
 //|
 
 //|     def __init__(self, *, width: int) -> None:
```
Update this too please.

gamblor21 replied:
I think I'm missing something for this (and the last) comment. The new description is what the class does or were you looking for something else? (the FrameBuffer wording was awkward so I updated it)

tannewt replied:
This function signature needs all of the arguments it takes below and documentation for each one.

gamblor21 replied:
Sorry and thanks. Totally was blanking out on what was missing last night.

tannewt replied:
No problem. :-)

tannewt commented on shared-bindings/is31fl3741/IS31FL3741.c:
```diff
@@ -124,185 +77,77 @@ STATIC mp_obj_t is31fl3741_IS31FL3741_deinit(mp_obj_t self_in) {
 
-//|     width: int
-//|     """The width of the display, in pixels"""
+//| def is31fl3741_write(mapping: Tuple[int, ...], buf: ReadableBuffer) -> None:
```
```suggestion
//| def write(mapping: Tuple[int, ...], buf: ReadableBuffer) -> None:
```

tannewt commented on shared-bindings/is31fl3741/IS31FL3741.c:
```diff
@@ -124,185 +77,77 @@ STATIC mp_obj_t is31fl3741_IS31FL3741_deinit(mp_obj_t self_in) {
-    { MP_ROM_QSTR(MP_QSTR_width), MP_ROM_PTR(&is31fl3741_IS31FL3741_width_obj) },
-    { MP_ROM_QSTR(MP_QSTR_height), MP_ROM_PTR(&is31fl3741_IS31FL3741_height_obj) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_write), (mp_obj_t)&is31fl3741_IS31FL3741_write_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_reset), (mp_obj_t)&is31fl3741_IS31FL3741_reset_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_enable), (mp_obj_t)&is31fl3741_IS31FL3741_enable_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_set_global_current), (mp_obj_t)&is31fl3741_IS31FL3741_set_global_current_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_set_led), (mp_obj_t)&is31fl3741_IS31FL3741_set_led_obj },
```
Document these please.

tannewt reviewed (changes requested):
Thank you for the reorg. I like the separation you have between the IS31FL3741 and FrameBuffer classes! I just requested some doc updates to match. Thanks!

tannewt commented on shared-bindings/is31fl3741/FrameBuffer.c:
```diff
@@ -0,0 +1,312 @@
+//| class IS31FL3741_FrameBuffer:
+//|     """Creates an in-memory framebuffer for a IS31FL3741 device."""
+//|
+//|     def __init__(self, is31: is31fl3741.IS31FL3741, width: int, height: int, mapping: Tuple[int, ...], *,
```
```suggestion
//|     def __init__(self, *, is31: is31fl3741.IS31FL3741, width: int, height: int, mapping: Tuple[int, ...],
```

tannewt commented on shared-bindings/is31fl3741/IS31FL3741.c:
```diff
@@ -124,185 +80,101 @@ STATIC mp_obj_t is31fl3741_IS31FL3741_deinit(mp_obj_t self_in) {
+//|     :param int value: value to set the LED to 0x00 to 0xFF
+//|     :param int page: page to write to 0 or 2. If the LED is a >= 180
+//|       the routine will automatically write to page 1 or 3 (instead
+//|       of 0 ot 2)"""
```
```suggestion
//|       of 0 or 2)"""
```

tannewt commented on shared-bindings/is31fl3741/FrameBuffer.c:
```diff
@@ -0,0 +1,312 @@
+STATIC mp_obj_t is31fl3741_FrameBuffer_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_is31, ARG_width, ARG_height, ARG_framebuffer, ARG_mapping, ARG_scale, ARG_gamma };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_is31, MP_ARG_OBJ | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
+        { MP_QSTR_width, MP_ARG_INT | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
+        { MP_QSTR_height, MP_ARG_INT | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
```
Are these actually required to be keywords? They are documented like position arguments and used in the library as such. Probably best to just remove the `MP_ARG_KW_ONLY` here.

gamblor21 replied:
I removed it from the mapping as well (also required) and reorged everything so it was in the right order.

tannewt reviewed (changes requested):
Very close! Just a typo and a question about kw_only required arguments.

gamblor21 commented:
Not sure why there was one cancelled check, but with the rest passing I don't think it is worth re-running the whole thing.

tannewt commented on shared-bindings/is31fl3741/IS31FL3741.c:
```diff
@@ -32,82 +32,38 @@
     static const mp_arg_t allowed_args[] = {
-        { MP_QSTR_width, MP_ARG_INT | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
-        { MP_QSTR_height, MP_ARG_INT | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
         { MP_QSTR_i2c, MP_ARG_OBJ | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
```
```suggestion
        { MP_QSTR_i2c, MP_ARG_OBJ | MP_ARG_REQUIRED },
```

tannewt commented on shared-bindings/is31fl3741/FrameBuffer.c:
```diff
@@ -0,0 +1,312 @@
+//| class IS31FL3741_FrameBuffer:
+//|     """Creates an in-memory framebuffer for a IS31FL3741 device."""
+//|
+//|     def __init__(self, *, is31: is31fl3741.IS31FL3741, width: int, height: int, mapping: Tuple[int, ...],
```
```suggestion
//|     def __init__(self, is31: is31fl3741.IS31FL3741, width: int, height: int, mapping: Tuple[int, ...], *,
```

tannewt reviewed (changes requested):
Sorry, two more minor things. I suggested conflicting changes to you before wrt the `*` and kw_only. More folks are using the stubs so think it's worth correcting now.

tannewt reviewed (changes requested):
Thank you! This looks good. It just needs conflicts resolved and then will be good to go.

gamblor21 commented:
All the merge fixes are good. One build failed with a 403 error that doesn't seem related to the changes.

tannewt commented on shared-module/is31fl3741/IS31FL3741.c:
```diff
@@ -248,92 +82,76 @@ void common_hal_displayio_is31fl3741_begin_transaction(is31fl3741_IS31FL3741_obj
+    common_hal_is31fl3741_set_page(self, 4);
     uint8_t gcur = 0x01; // global current command
+    common_hal_busio_i2c_write(self->i2c, self->device_address, &gcur, 1);
+
     uint8_t data = 0;
-    common_hal_busio_i2c_write_read(i2c, addr, &gcur, 1, &data, 1);
+    common_hal_busio_i2c_read(self->i2c, self->device_address, &data, 1);
```
It looks like this should be a `write_read`.

dhalbert replied:
and remove line 131, I think.

gamblor21 replied:
The latest build should have removed both and replace them with the `write_read`

tannewt reviewed (changes requested):
One spot where I think you want a `write_read`.

tannewt reviewed (approved):
Looks good to me! Thank you!
